src/api/c/random_engine.cpp

```diff
@@ -0,0 +1,357 @@
+                    e->mask,
+                    getArray<uint>(e->recursion_table),
+                    getArray<uint>(e->temper_table),
+                    getArray<uint>(e->state)));
+    } else {
+        return getHandle(normalDistribution<T>(dims, e->type, e->seed, e->counter));
+    }
+}
+
+af_err af_create_random_engine(af_random_engine *engineHandle, af_random_type rtype, uintl seed)
+{
+    try {
+        RandomEngine e;
+        e.type = rtype;
+        e.seed = seed;
+        e.counter = 0;
+        if (rtype == AF_RANDOM_MERSENNE_GP11213) {
+            AF_CHECK(af_create_array(&e.pos, pos, 1, &MaxBlocks, u32));
+            AF_CHECK(af_create_array(&e.sh1, sh1, 1, &MaxBlocks, u32));
+            AF_CHECK(af_create_array(&e.sh2, sh2, 1, &MaxBlocks, u32));
+            e.mask = mask;
+            AF_CHECK(af_create_array(&e.recursion_table, recursion_tbl, 1, &TableLength, u32));
+            AF_CHECK(af_create_array(&e.temper_table, temper_tbl, 1, &TableLength, u32));
+            e.state = getHandle(initMersenneState(seed, getArray<uint>(e.recursion_table)));
+        }
+        *engineHandle = getRandomEngineHandle(e);
+    } CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+af_err af_retain_random_engine(af_random_engine *outHandle, const af_random_engine engineHandle)
+{
+    try {
+
+        RandomEngine engine = *(getRandomEngine(engineHandle));
+        RandomEngine out;
+
+        out.type = engine.type;
+        out.seed = engine.seed;
+        out.counter = engine.counter;
+        AF_CHECK(af_retain_array(&out.pos, engine.pos));
+        AF_CHECK(af_retain_array(&out.sh1, engine.sh1));
+        AF_CHECK(af_retain_array(&out.sh2, engine.sh2));
+        out.mask = engine.mask;
+        AF_CHECK(af_retain_array(&out.recursion_table, engine.recursion_table));
+        AF_CHECK(af_retain_array(&out.temper_table, engine.temper_table));
+        AF_CHECK(af_retain_array(&out.state, engine.state));
+
+        *outHandle = getRandomEngineHandle(out);
+
+    } CATCHALL;
+    return AF_SUCCESS;
+}
+
+af_err af_random_engine_set_type(af_random_engine *engine, const af_random_type rtype)
+{
+    try {
+        RandomEngine e = *(getRandomEngine(engine));
+        if (rtype != e.type) {
+            bool empty;
+            if (rtype == AF_RANDOM_MERSENNE_GP11213) {
+                af_is_empty(&empty, e.state);
+                if (empty) {
+                    AF_CHECK(af_create_array(&e.pos, pos, 1, &MaxBlocks, u32));
+                    AF_CHECK(af_create_array(&e.sh1, sh1, 1, &MaxBlocks, u32));
+                    AF_CHECK(af_create_array(&e.sh2, sh2, 1, &MaxBlocks, u32));
+                    e.mask = mask;
+                    AF_CHECK(af_create_array(&e.recursion_table, recursion_tbl, 1, &TableLength, u32));
+                    AF_CHECK(af_create_array(&e.temper_table, temper_tbl, 1, &TableLength, u32));
+                    e.state = getHandle(initMersenneState(e.seed, getArray<uint>(e.recursion_table)));
+                }
```

Comment:
Don't you need to do these calls if its not `AF_RANDOM_MERSENNE_GP11213`? Does the random engine stay valid for other enums that don't need these calls?